extensions/wikia/Rail/RailController.class.php

```diff
@@ -8,139 +8,159 @@ class RailController extends WikiaController {
 
-			$this->css = $sassFiles = [];
-			foreach (array_keys($this->app->wg->Out->styles) as $style) {
-				if ($wgAllInOne && $assetManager->isSassUrl($style)) {
-					$sassFiles[] = $style;
-				} else {
-					$this->css[] = $style;
-				}
-			}
+		foreach ( $railModules as $railModule ) {
+			$this->railLazyContent .= $this->app->renderView(
+				$railModule[0], /* Controller */
+				$railModule[1], /* Method */
+				$railModule[2] /* array of params */
+			);
+		}
 
-			if (!empty($sassFiles)) {
-				$excludeScss = (array) $this->getRequest()->getVal('excludeScss', []);
-				$sassFilePath = (array) $assetManager->getSassFilePath($sassFiles);
-				$includeScss = array_diff($sassFilePath, $excludeScss);
+		// ad mix experiment uses a wrapper to group recirculation and ad placeholder
+		if ( !$isAdMixExperimentEnabled ) {
+			$this->railLazyContent .= Html::element( 'div', [ 'id' => 'WikiaAdInContentPlaceHolder' ] );
+		}
 
-				// SUS-771: Log any duplicate CSS that rail modules try to load but are already loaded by Oasis skin
-				$duplicateScss = array_intersect( $sassFilePath, $excludeScss );
-				if ( count( $duplicateScss ) ) {
-					Wikia\Logger\WikiaLogger::instance()->info( 'SUS-771', [
-						'styles' => json_encode( $duplicateScss )
-					] );
-				}
+		$this->css = $sassFiles = [];
+		foreach ( array_keys( $this->app->wg->Out->styles ) as $style ) {
+			if ( $wgAllInOne && $assetManager->isSassUrl( $style ) ) {
+				$sassFiles[] = $style;
+			} else {
+				$this->css[] = $style;
+			}
+		}
 
-				if (!empty($includeScss)) {
-					$this->css[] = $assetManager->getSassesUrl($includeScss);
-				}
+		if ( !empty( $sassFiles ) ) {
+			$excludeScss = (array) $this->getRequest()->getVal( 'excludeScss', [] );
+			$sassFilePath = (array) $assetManager->getSassFilePath( $sassFiles );
+			$includeScss = array_diff( $sassFilePath, $excludeScss );
+
+			// SUS-771: Log any duplicate CSS that rail modules try to load but are already loaded by Oasis skin
+			$duplicateScss = array_intersect( $sassFilePath, $excludeScss );
+			if ( count( $duplicateScss ) ) {
+				Wikia\Logger\WikiaLogger::instance()->info(
+					'SUS-771',
+					[
+						'styles' => json_encode( $duplicateScss )
+					]
+				);
 			}
 
-			// Do not load user and site jses as they are already loaded and can break page
-			$oldWgUseSiteJs = $wgUseSiteJs;
-			$oldWgAllowUserJs = $wgAllowUserJs;
-			$wgUseSiteJs = false;
-			$wgAllowUserJs = false;
+			if ( !empty( $includeScss ) ) {
+				$this->css[] = $assetManager->getSassesUrl( $includeScss );
+			}
+		}
 
-			$this->js = $this->app->wg->Out->getBottomScripts();
+		// Do not load user and site jses as they are already loaded and can break page
+		$oldWgUseSiteJs = $wgUseSiteJs;
+		$oldWgAllowUserJs = $wgAllowUserJs;
+		$wgUseSiteJs = false;
+		$wgAllowUserJs = false;
 
-			$wgUseSiteJs = $oldWgUseSiteJs;
-			$wgAllowUserJs = $oldWgAllowUserJs;
-			$wgTitle = $oldWgTitle;
-		}
+		$this->js = $this->app->wg->Out->getBottomScripts();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this code to rely on RequestContext instead of context-dependent globals. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)